docs/notebooks/flax_basics.ipynb

```diff
@@ -0,0 +1,911 @@
+{
```

Comment:
<p>Could you explain why the two names <code>counter</code> and <code>count</code> ?</p><p> </p><p>Maybe it would make more sense to use a "standard" name for the collection name? Do we want to motivate users to come up with their own collections?</p><p> </p><p>I would find it more straightforward to do an example with a real application for state (like batchnorm) to make clear what collections are useful for. Creating a Counter as a module feels a bit contrived.</p><p> </p><p>WDYT about using the same name for the value returned by <code>model.init()</code> everywhere? I would vote for simply <code>variables</code> (though pydoc says <code>data</code>, note that in the example above you use <code>params</code>  for the returned values) and then use for <code>y, mutable_variables = model.apply(variables, x, mutable=['counter'])</code> and <code>optimizer_def.create(variables['params'])</code> - i.e. <code>variables</code> instead of <code>params</code> (in the example above), and never using the <code>init_variables</code> (I don't see why we need to give them another name, makes it very easy to forget deleting them). Calling the update variables <code>state</code>  (or <code>updated_state</code>)  makes it less clear that the selection there is directly determined by the argument <code>mutable=</code>. <span class="ql-cursor">﻿</span></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/flax/pull/542/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

Reply:
<p>I see, so I think I miss an explanation of the collections and params, I'll add something.</p><p> </p><p>I don't have an opinion on having standard collection names, I guess beyond the batch norm one it's hard to think about all the use cases. This shouldn't probably be part of the best practices for Flax.</p><p> </p><p>On the contrived nature of the example I agree, and I actually hesitated between this and using batchnorm directly. The issue I see with BatchNorm is that I wanted to have the full implementation here (and not too long), so that you could see how to use variables. Problem is that a proper batch norm is a bit too long (and not everyone knows on top of their head what BatchNorm exactly is). However, I think it would be good to have something in between, but I didn't have a better idea at this point. If you have one happy to change to something that makes more sense indeed!</p><p> </p><p>About the names yeah I see, actually I changed the naming for the very specific reasons you mentioned, but ended up also confusing indeed. Will change across the doc to make it clearer and more consistent</p><p> </p>

Reply:
What about using a "simplified version of batchnorm", e.g. simply keeping the running average and subtracting that from the inputs? That should be fairly short and you could say that IRL you would ues batchnorm & link to the Flax implementation for reference.

As for naming : Note that currently you say `print('updated variables:\n', updated_state)` (i.e. `variables` but then only print `state` which is a subset of the variables).

Reply:
<p>Makes sense, let me revamp a bit that section and address those.</p>